core/src/main/java/org/eclipse/hono/util/EncodedPassword.java

```diff
@@ -0,0 +1,170 @@
+     * @return Encoded password
+     * @throws IllegalArgumentException if candidate hashed-password secret does not contain a password hash
+     */
+    public static EncodedPassword fromHonoSecret(final JsonObject honoSecret) throws IllegalArgumentException {
+        Objects.requireNonNull(honoSecret);
+
+        final String pwdHash = honoSecret.getString(CredentialsConstants.FIELD_SECRETS_PWD_HASH);
+        if (pwdHash == null) {
+            throw new IllegalArgumentException("candidate hashed-password secret does not contain a pwd hash");
+        }
+
+        final String hashFunction = honoSecret.getString(
+                CredentialsConstants.FIELD_SECRETS_HASH_FUNCTION,
+                CredentialsConstants.DEFAULT_HASH_FUNCTION);
+
+        final String encodedSalt = honoSecret.getString(CredentialsConstants.FIELD_SECRETS_SALT);
+        final EncodedPassword encodedPassword = new EncodedPassword();
+        encodedPassword.hashFunction = hashFunction;
+        encodedPassword.passwordHash = pwdHash;
+        if (encodedSalt != null) {
+            encodedPassword.salt = Base64.getDecoder().decode(encodedSalt);
+        }
+        return encodedPassword;
+    }
+
+    /**
+     * Creates a new instance from the String in {salt}password format. The password can be in raw or hashed form.
+     * This method is used by Spring Security password encoders, to parse submitted values.
+     *
+     * @param formattedPassword Password in the {salt}password format
+     * @param raw {@true} if password is in raw form, {@false} if it's hashed
+     * @return Encoded password
+     */
+    public static EncodedPassword parse(final String formattedPassword, final boolean raw) {
+        Objects.requireNonNull(formattedPassword);
+
+        final EncodedPassword encodedPassword = new EncodedPassword();
+
+        final int start = formattedPassword.indexOf(PREFIX);
+        String parsedPassword = formattedPassword;
+
+        if (start == 0) {
+            final int end = formattedPassword.indexOf(SUFFIX, start);
+            if (end > 0) {
+                encodedPassword.salt = Base64.getDecoder().decode(formattedPassword.substring(start + 1, end));
+                parsedPassword = formattedPassword.substring(end + 1);
+            }
+        }
+
+        if (raw) {
+            encodedPassword.rawPassword = parsedPassword;
+        } else {
+            encodedPassword.passwordHash = parsedPassword;
+        }
+
+        return encodedPassword;
+    }
+
+    /**
+     * Encode the password based on the provided hash function and salt.
+     *
+     * @param hashFunction Hash function used to encode this password
+     * @param salt Salt in the form of byte array
+     * @param password Password to be encoded
+     * @return Encoded password as a Base64 formatted String
+     */
+    public static String encode(final String hashFunction, final byte[] salt, final String password) {
```

Comment:
I have no clue what the result of this method is. Please document what the returned string looks like and how it is is created, otherwise nobody (including me) will ever understand what this method does...

Reply:
I added some more description, but I don't now what else I can write (it's there in the "@return" description). A simple Base64 encoded password hash.

Reply:
The point is that there is no such thing as *a simple Base64 encoded password hash* because the result mainly depends on the input bytes that you apply the hash function to. So, if I do not know what how the hash has been computed, I have no clue what I am supposed to do with the result returned by this method.